File: src/part3/code3-6.py
if __name__ == "__main__":
    N: int = int(input())
    W: int = int(input())

    a = []
    for _ in range(N): a.append(int(input()))

    # bitは2^N通りの部分集合全体を動きます
    exist: bool = False
    for bit in range(1 << N):
        sum: int = 0 # 部分集合が含まれる要素の和
        for i in range(N):
            # i番目に要素a[i]が部分集合に含まれているかどうか
            if bit & (1 << i):
                sum += a[i]

        if sum == W: exist = True

    if exist: print("Yes")
    else: print("No")

File: src/part3/code3-6.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int N, W;
    cin >> N >> W;
    vector<int> a(N);
    for (int i = 0; i < N; ++i) cin >> a[i];

    // bitは2^N通りの部分集合全体を動く
    bool exist = false;
    for (int bit = 0; bit < (1 << N); ++bit) {
        int sum = 0; // 部分集合に含まれているかどうか
        for (int i = 0; i < N; ++i) {
            if (bit & (1 << i)) {
                sum += a[i];
            }
        }
        // sumがWに一致するかどうか
        if (sum == W) exist = true;
    }

    if (exist) cout << "Yes" << endl;
    else cout << "No" << endl;
}